libs/defi/shared/src/components/Swapper/Swapper.tsx

```diff
@@ -134,91 +127,76 @@ export const Swapper = ({
       }}
       onApproveFailure={(state) => {
         const { error, trackId } = state;
-        updateActivity({
-          ...state,
+        const updated = updateActivity({
           id: trackId,
           status: 'error',
           error: formatError(error),
         });
-        pushNotification({
-          icon: <TokenIcon token={state.tokenIn} />,
-          title: intl.formatMessage({
-            defaultMessage: 'Error while approving',
-          }),
-          message: formatError(error),
-          severity: 'error',
-        });
+        if (updated) {
+          pushNotification({
```

Comment:
`pushNotificationForActivity` would remove the need to understand the intricacies if pushing a notification which is entirely dependent on the activity.

Reply:
I don't get your point, we push a notification from the place "orchestrating" the behaviour which seems the logical place for it. I don't see how `pushNotificationForActivity` would help where it would become a giant if-else to account for all different steps. And as we discussed, I'm really against dispatching notifications from activity lifecycle, this would introduce dependency between the two systems which doesn't really make sense.